docs/RFCs/19042021_Cert_Management_In_CRDB_Helm_Chart.md

```diff
@@ -0,0 +1,243 @@
+       - name: NODE_CERT_DURATION
+         value: {{ default 365 .Values.tls.certs.generate.nodeCertDuration}}
+       - name: Client_CERT_DURATION
+         value: {{ default 365 .Values.tls.certs.generate.clientCertDuration}}
+       {{- if and (tls.certs.generate.caProvided .Values.tls.certs.generate.caSecret) }}
+           {{- if not (lookup "v1" "Secret" ".Release.Namespace" ".Values.tls.certs.generate.caSecret")}}
+           {{ fail "CA secret doesn't exist in cluster"}}
+           {{- end }}
+       - name: CA_SECRET
+         values:
+       {{- end }}
+    ```
+ 
+* 3 empty secret will be created in Helm chart for `cockroachdb-ca`, `cockroachdb-node` and `cockroachdb-root` if `tls.certs.generate.enabled`
+ is set.
+  * Data to these secrets will be populated in the `pre-install` job.
+  * In case CA is provided by the user, then `cockroachdb-ca` secret is skipped.
+  * Annotation is set on all the secrets created by CockroachDB; eg: `managed-by: cockroachdb`
+ 
+* A cronjob will be created in Helm chart when `tls.certs.generate.rotateCerts` is set.
+  * This cronjob will run periodically to rotate the certificates.
+  * The schedule of the cronjob will be of minimum `clientCertDuration` and `nodeCertDuration` minus some delta.
+  * On every scheduled run, it will check if there is any certificate that is going to expire before the next scheduled run,
+   if yes then it will renew the certificates.
+  
+  * <b>The cronjob will use the same `pre-install` job image for certificate rotations. The `pre-install` job image binary will have an argument `--rotate` for handling certificate rotation.</b>
+ 
+* The Stateful pod will be changed to only run `copy-certs` initContainer to copy the certificates from Node secret to emptyDir volume.
+  The rest of the main DB container flow will remain the same.
+* The post-install job will be changed to only run `copy-certs` initContainer to copy the certificates from Client secret to emptyDir volume.
+  The rest of the main cluster-init container flow will remain the same.
+ 
+`TODO: Need to identify how to generate the SIGHUP signal in all the Nodes for certificate renewal`
+ 
+### Certificate Management Service Implementation
+ 
+ 
+### Overall Flow
+ 
+* Check if CA secret is empty
+  * if yes, [Generate-CA](#generate-ca), [Generate-Node-Cert](#generate-node-cert) and [Generate-Client-Cert](#generate-client-cert); return
+  * if not, [Validate Secret Annotations for CA](#validate-secret-annotations)
+    * if valid, CA is intact
+    * if not, [Annotate-CA](#annotate-ca), [Generate-Node-Cert](#generate-node-cert) and [Generate-Client-Cert](#generate-client-cert); return
+* Check if [CA requires rotation](#check-cert-for-regeneration)
+   * if yes, follow [Rotate-CA](#rotate-ca); return
+* Check if Node or Client certificates [needs to be regenerated](#check-cert-for-regeneration):
+   * if yes, follow [Generate-Node-Cert](#generate-node-cert) and [Generate-Client-Cert](#generate-client-cert)
```

Comment:
We can add the retries on the job failure. Also, Instead of spatting the failures in logs only, we can add the custom events on the cronJob about each operation we are performing. This would give user a better understanding of the cronjob when user runs `kubectl describe cronjob`